### ISSUE TITLE
IOS Game Stucks & Freezes

### ISSUE THREAD

**Issue by yeaaqbi** [Scirra/Construct-3-bugs#2553]
## Problem description
I have exported my game to android and IOS, it works perfect on Android but it is some times freezes on iphone! I have debugged the game on XCode and noticed that it gives me the following error.

## Attach a .c3p

[‎⁨أم الطنافس الفوقا⁩.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/2799957/c3p.zip)

## Steps to reproduce

1. Export the project to XCode project
2. Connect your iphone to mac and run the game from XCode to the connected iphone
3. Case 1: Move around the game, play with it go to other layouts ... etc. then the game stucks and gives the attached error in the xcode debug console
   Case 2: close the game and reopen it on your iphone, there is a black screen stucks and also giving the same error. 

## Observed result

The game stucks or gives me black screen, the game runs after re open it many times

## Expected result

the game should run without any problem - at least without any stuck or black screen
the game should works fine when open it, not reopen it many times to make it runs fine
## First affected release
r.133

## Affected browsers/platforms

* Chrome: no
* Firefox: no
* Edge: no
* Safari: yes

## System details
Iphone 7 
IOS 12.1
C3 realease: r 135.2

## Errors 

<img width="426" alt="51671318-53703800-1fd1-11e9-9867-befb0ea61495" src="https://user-images.githubusercontent.com/3922086/51800083-0db2aa00-2233-11e9-8bb3-f983aa9e83d9.png">

**Comment by Gamedevolop** [comment]
@yeaaqbi  it looks exactly the same bug that I reported a few days ago, where on PhoneGap fails the Builds or  on Xcode Freezes when trying to enter into the Levels and shows me the same Error message as you 
[Bug](https://github.com/Scirra/Construct-3-bugs/issues/2548)

**Comment by yeaaqbi** [comment]
@Gamedevolop the problem is that I want to go live this Friday, I hope the fix will be easier than I think :(

**Comment by AshleyScirra** [comment]
Please see our bug report guidelines. The project uses a third-party addon "Photon Realtime by Chadori_RebornXD". To prove that this plugin is not causing the problem, please remove it from the project and test again, and provide that project file if the issue still does reproduce. At the very minimum, at least tick "bundle addons" in project properties so that the specific version of the addon you are using is embedded in to the project, so we do not have to try to hunt down the addon and the matching version that you happen to be using.

FWIW this might be an issue with unicode since the filenames appear to be in Arabic (especially since it's a right-to-left language). Using English (ASCII) names for objects might work around the issue, and if it does, proves that it's related to unicode filenames.

**Comment by yeaaqbi** [comment]
@AshleyScirra  I will try to remove it, changing the name to english and back to you. Thanks a lot, wait my feedback after checking.

**Comment by yeaaqbi** [comment]
@AshleyScirra I have tried again, it is the same issue. I am attaching the project without the photon plugin or the arabic title. 
[Tanafis.c3p.zip](https://github.com/Scirra/Construct-3-bugs/files/2803286/Tanafis.c3p.zip)

**Comment by sjrya** [comment]
@AshleyScirra 

Here is what I'm getting in the safari debugger:

`[Info] [WebGL] Shader program '<default>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<default-device-transform>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<point>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_351d81300000005' to match BindAttributeLocation request.WARNING: Could not fin…" (cordova.js, line 1732)
[Info] [WebGL] Shader program '<fill>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request.WARNING: Output of verte…" (cordova.js, line 1732)
[Info] [WebGL] Shader program '<lineargradient>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<hardellipse>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<hardellipseoutline>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<smoothellipse>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<smoothellipseoutline>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<smoothline>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] [WebGL] Shader program '<tilemap>' compilation log:  – "WARNING: Could not find vertex shader attribute 'webgl_359fbc7d00000008' to match BindAttributeLocation request." (cordova.js, line 1732)
[Info] Made with Construct 3, the game and app creator :: https://www.construct.net (cordova.js, line 1732)
[Info] [C3 runtime] Hosted in DOM, rendering with WebGL 1 [Apple A10 GPU] (cordova.js, line 1732)
[Log] [C3 IAP] – "Initialising" (cordova.js, line 1732)`

**Comment by AshleyScirra** [comment]
That log does not show any serious problems at all. So unfortunately we're none the wiser yet.

**Comment by sjrya** [comment]
I'm also receiving 'Blobresource error 1' messages now. 

Could it be something to do with the webkit? I've been trying to look on other trackers to see if people are getting similar issues (some seem to be) - but I'm really not too sure what I'm looking for!

**Comment by Boulerzzz** [comment]
feel your pain @sjrya 

I've tried literally everything the last few days. I wish I knew why it happens, or at least could help the devs with this. Its so strange. You can switch layout 10-100 times and nothing happens but then the game all of a sudden freezes. No input or nothing works anymore. At first I though it froze after the layout changed. But then I noticed it never actually changes layout. It just freezes when the "go to layout" action triggers. Restart layout seems to be fine. It is just the go to layout action.
I exported like 10 versions of my game to testflight (all with small diffrences to the way it changes layout) and then I entered and exited levels 1000 times on every version on my iphone 6s.
I did see that the freezes were less frequent when you set spritesheet size to 4k. But my sample of 1000 tries is really small so it may just be coincidence. (but maybe its something with the spritesheets?)

We probably need to go back to c2 runtime but I'd hate to redo so much work in order to replace behaviors and expressions not compatible. :((((

**Comment by yeaaqbi** [comment]
Hi @AshleyScirra , 

Hope you are good, any updates about this bug plz?

**Comment by Boulerzzz** [comment]
@yeaaqbi does your game only freeze when the "go to layout xxxxx" action is called? 

That is the only action that freeze my game. How often does it happen to you?

**Comment by Ninjadoodle** [comment]
Hi @AshleyScirra - I'm getting exactly the same thing with my iOS Cordova (xCode) export. My game works perfectly 99% of the time, then very rarely when tapping the menu play button, it fades out and never goes to the next scene. You will literally need to sit there and test the game over and over again for it to happen. I'm using r136 stable and only official C3 plugins.

**Comment by AshleyScirra** [comment]
I would guess this is a WebKit/iOS bug of some kind, or maybe an issue with Cordova... there also seems to be conflicting information in this report: in one case the error is "Failed to load texture", and in another "Blobresource error 1". These may be unrelated, but we can't tell yet due to the lack of good information about this issue.

Does this only happen in the C3 runtime or does it affect both the C2 and C3 runtimes?

**Comment by Boulerzzz** [comment]
@AshleyScirra 
I changed my game to C2 runtime and everything works great now.

**Comment by Ninjadoodle** [comment]
Hi @AshleyScirra - I'm using C3 runtime, because I use the new tween plugin etc. For me it seems to happen on scene change.

**Comment by Ninjadoodle** [comment]
@AshleyScirra - anything we can do or provide, to make finding the issue easier? Makes it hard as I can't really send an xcode project with the correct info.

**Comment by sjrya** [comment]
Hey. 

I managed to fix my issue but I have no idea if we’re experiencing the same thing.

-I tidied up all my sprites and made them more mobile efficient as I thought maybe it was causing a memory crash 

-After some trial and error, I converted all my audio files (Sounds) back to .m4a from .webm and that seems to have solved my problem.

Again, not sure if this will help you guys or I just got lucky. WebKit doesn’t seem to be loving .webms for me atm.
So maybe if you guys have any sounds, you could try that

**Comment by RichardAstrom** [comment]
Hi! @AshleyScirra I also get the "Failed to load texture" error in XCode. I can keep on playing my game but some sprites are not visible. They are there though because I can interact with them (for example sometimes the start button disappears but I can still press it).

The problem is that this only happens occasionally and seems to be totally random. Unfortunately when it happens, it totally breaks the game.

I'm using r128.3 beta with the C3 engine and the Construct build service to Xcode project with Xcode 10.1. I have also tried to build the game and installed via test flight and it still happens.

![failedtoloadtexture](https://user-images.githubusercontent.com/3534495/53333559-81110f80-38f6-11e9-946d-bed16322f919.png)

**Comment by AshleyScirra** [comment]
> I'm using r128.3 beta

Does it reproduce in the latest release?

**Comment by RichardAstrom** [comment]
Yes! Sorry I wrote the wrong release number :) I'm using r138.2.

**Comment by AshleyScirra** [comment]
There's a small tweak in r139 to make the C3 runtime handle Cordova file reads more similarly to how the C2 runtime does, since it seems to work fine there. Does it make a difference for anyone?

**Comment by RichardAstrom** [comment]
I tried r139 and after some testing I got the same error again. You can see the error message in the bottom of the attached screenshot.

I also tried to build the game with Cordova CLI and the latest iOS 5.0.0 and so far I haven't been able to reproduce "Failed to load texture". But I guess I'll have to test it some more :)

![image](https://user-images.githubusercontent.com/3534495/53342500-5cc12d00-390e-11e9-921e-c851eea2313a.png)

**Comment by Ninjadoodle** [comment]
@AshleyScirra Tried using r139 just now, and still getting the same freezing on scene change.

**Comment by RichardAstrom** [comment]
I get the error on 139 using the Construct build service and when I've built the game with Cordova CLI as well unfortunately. "Failed to load texture". Right now it's impossible for us to release our game.

**Comment by yeaaqbi** [comment]
Me too the same, I am waiting the fix for more than month and as I am a developer I know how they can not get the real issue yet. Hope they will fix it soon. I had a release too weeks a go and I was lucky that I started to develop the game in parallel with unity 3d. So I was really lucky. But also I am waiting the fix since I really like how C3 looks now ! it is really wonderful !!

**Comment by shortercode** [comment]
I think I have tracked down the source of this, it appears as though webkit is prematurely releasing file references. This causes non deterministic breakages when we need to load a lot of stuff, the more files and the less RAM a device has the quicker webkit will attempt to release them and the more likely it is to break. In the example project it would occasionally fail starting the game for me, and nearly always fail moving into the first level for example.

With that in mind I believe I have a solution that fixes this, just by rearranging how the file loader works. Following some additional testing ( to see if my explanation holds true and the fix works correctly ) I will commit this patch for the next release.

**Comment by yeaaqbi** [comment]
@shortercode so excited for the next release !! Thanks  a lot ;)

**Comment by Ninjadoodle** [comment]
Awesome, looking forward to testing it out :) Thank you!

**Comment by HRTStudio** [comment]
@AshleyScirra 

Construct r139
Facing the same issue when exported for iOS in Xcode version. 
Helishooter[2250:394476] ERROR: Failed to load texture:  [object Event]
and in android it shows black screen on device when exported.

**Comment by AshleyScirra** [comment]
r140 is now out, please check again!

**Comment by yeaaqbi** [comment]
@AshleyScirra After exporting the project and opened it in XCode, it gives me the errors in the attached image, is it something related to C3 runtime? it appeared after I have added the local storage to the app to read values. 
![screen shot 2019-03-04 at 11 03 33 pm](https://user-images.githubusercontent.com/3922086/53762948-1cdfe400-3ed2-11e9-9623-b34be6a35228.jpg)

**Comment by yeaaqbi** [comment]
@AshleyScirra the errors disappears when I have removed the Mobile Advert plugin, so it is not related to the Local Storage plgin

**Comment by yeaaqbi** [comment]
@AshleyScirra it seems the app works without freezing now ! Thanks a lot for hard work guys <3 I will keep testing and transferring between the layouts to see if anything happens.

**Comment by AshleyScirra** [comment]
Thanks for confirming. If you have any other problems please file a new issue.
